app/controllers/coronavirus_form/address_lookup_controller.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class CoronavirusForm::AddressLookupController < ApplicationController
+  def submit
+    redirect_to check_your_answers_url if session[:check_answers_seen]
+
+    if params[:postcode].blank?
+      flash.now[:validation] = [{ text: I18n.t("coronavirus_form.questions.address_lookup.hint"), field: "postcode" }]
+      render controller_path, status: :unprocessable_entity
+    elsif params[:address_string].present?
+      update_params
+      redirect_to support_address_path
+    else
+      begin
+        results = helpers.postcode_lookup(params[:postcode])
+        @addresses = generate_address_strings(results)
+        @postcode = params[:postcode]
+        render controller_path, status: :ok
+      rescue AddressLookupError => e
+        flash.now[:validation] = [{ text: e.message, field: "postcode" }]
+        render controller_path, status: :unprocessable_entity
+      end
+    end
+  end
+
+private
+
+  def update_params
+    uprn_and_address = params[:address_string].split("|")
```

Comment:
If you use JSON in the input values you could then parse that as JSON here, rather than parsing a string.

E.g. 

```html
<select>
   <option value='{"uprn":"1234","postcode":"abc","address_line_1":"10 Downi..."..}'>
      10 Downing Street, London..
    </option>
</select>
```

```ruby
address = validate_address(JSON.parse(selected_address))
```

This should make it easier to parse and validate the selected address.

Reply:
In order to make this work we had to make another API call to the uprn end-point.  We are not using JSON, but we are using all of the fields directly from the API response, so strings are no longer a problem.